SimCalorimetry/HGCalSimProducers/plugins/PreMixingHGCalWorker.cc

```diff
@@ -0,0 +1,83 @@
+#include "FWCore/Utilities/interface/RandomNumberGenerator.h"
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/HGCDigi/interface/PHGCSimAccumulator.h"
+#include "SimCalorimetry/HGCalSimProducers/interface/HGCDigitizer.h"
+
+#include "SimGeneral/PreMixingModule/interface/PreMixingWorker.h"
+#include "SimGeneral/PreMixingModule/interface/PreMixingWorkerFactory.h"
+
+class PreMixingHGCalWorker: public PreMixingWorker {
+public:
+  PreMixingHGCalWorker(const edm::ParameterSet& ps, edm::ProducerBase& producer, edm::ConsumesCollector&& iC);
+  ~PreMixingHGCalWorker() override = default;
+
+  PreMixingHGCalWorker(const PreMixingHGCalWorker&) = delete;
+  PreMixingHGCalWorker& operator=(const PreMixingHGCalWorker&) = delete;
+
+  void beginRun(const edm::Run& run, const edm::EventSetup& ES) override;
+  void endRun() override;
+  void initializeEvent(const edm::Event &e, const edm::EventSetup& ES) override {}
+  void addSignals(const edm::Event &e, const edm::EventSetup& ES) override;
+  void addPileups(const PileUpEventPrincipal&, const edm::EventSetup& ES) override;
+  void put(edm::Event &e,const edm::EventSetup& ES, std::vector<PileupSummaryInfo> const& ps, int bs) override;
+
+private:
+  edm::EDGetTokenT<PHGCSimAccumulator> signalToken_;
+
+  edm::InputTag pileInputTag_;
+
+  HGCDigitizer digitizer_;
+};
+
+PreMixingHGCalWorker::PreMixingHGCalWorker(const edm::ParameterSet& ps, edm::ProducerBase& producer, edm::ConsumesCollector&& iC):
+  signalToken_(iC.consumes<PHGCSimAccumulator>(ps.getParameter<edm::InputTag>("digiTagSig"))),
+  pileInputTag_(ps.getParameter<edm::InputTag>("pileInputTag")),
+  digitizer_(ps, iC)
+{
+  if(digitizer_.producesEEDigis()) {
+    producer.produces<HGCEEDigiCollection>(digitizer_.digiCollection());
+  }
+  if(digitizer_.producesHEfrontDigis()) {
+    producer.produces<HGCHEDigiCollection>(digitizer_.digiCollection());
+  }
+  if(digitizer_.producesHEbackDigis()) {
+    producer.produces<HGCBHDigiCollection>(digitizer_.digiCollection());
+  }
+}
+
+void PreMixingHGCalWorker::beginRun(const edm::Run& run, const edm::EventSetup& ES) {
+  digitizer_.beginRun(ES);
+}
+
+void PreMixingHGCalWorker::endRun() {
+  digitizer_.endRun();
+}
+
+void PreMixingHGCalWorker::addSignals(const edm::Event &e,const edm::EventSetup& ES) {
+  edm::Handle<PHGCSimAccumulator> handle;
+  e.getByToken(signalToken_, handle);
+  digitizer_.accumulate(*handle);
+}
+
+void PreMixingHGCalWorker::addPileups(const PileUpEventPrincipal& pep, const edm::EventSetup& ES) {
+  edm::Handle<PHGCSimAccumulator> handle;
+  pep.getByLabel(pileInputTag_, handle);
+  digitizer_.accumulate(*handle);
+}
+
+void PreMixingHGCalWorker::put(edm::Event &e,const edm::EventSetup& ES, std::vector<PileupSummaryInfo> const& ps, int bs) {
+  edm::Service<edm::RandomNumberGenerator> rng;
+  digitizer_.finalizeEvent(e, ES, &rng->getEngine(e.streamID()));
+}
+
+DEFINE_EDM_PLUGIN(PreMixingWorkerFactory, PreMixingHGCalWorker, "PreMixingHGCalWorker");
```

Comment:
it might be useful to define a separate macro for `DEFINE_EDM_PLUGIN(PreMixingWorkerFactory,...)`

Reply:
I was first thinking to object on the grounds of not saving that many keystrokes and me finding the `MixingModule` solution of https://github.com/cms-sw/cmssw/blob/master/SimGeneral/MixingModule/interface/DigiAccumulatorMixModFactory.h confusing, but then realized that a separate macro along 
```cpp
#define DEFINE_PREMIXING_WORKER(TYPE) DEFINE_EDM_PLUGIN(PreMixingWorkerFactory, TYPE, #TYPE) 
```
would allow removing the repetition of `PreMixingHGCalWorker, "PreMixingHGCalWorker"` (by forcing the C++ type and the plugin name to be the same).

Since such a change (when deployed fully) would touch all other `PreMixingWorker`s, do you have preference how to add that
* Add the macro and deploy fully in this PR
* Add the macro but deploy only the new PreMixingWorkers in this PR, deploy for the rest in a subsequent PR
* Implement in a subsequent PR
?

Reply:
Yes, that's a benefit of a separate macro. Let's do it all together in a subsequent PR.